src/segmentTemplate.js

```diff
@@ -0,0 +1,296 @@
+ * @param {number} values.Time
+ *        Timestamp value of the corresponding segment
+ * @return {string}
+ *         Segment url with identifiers replaced
+ */
+export const constructTemplateUrl = (url, values) =>
+  url.replace(identifierPattern, identifierReplacement(values));
+
+/**
+ * Uses information provided by SegmentTemplate@duration attribute to determine segment
+ * timing and duration
+ *
+ * @param {number} start
+ *        The start number for the first segment of this period
+ * @param {number} timeline
+ *        The timeline (period index) for the first segment of this period
+ * @param {number} timescale
+ *        The timescale for the timestamps contained within the media content
+ * @param {number} duration
+ *        Duration of each segment
+ * @param {number} sourceDuration
+ *        Duration of the entire Media Presentation
+ * @return {{number: number, duration: number, time: number, timeline: number}[]}
+ *         List of Objects with segment timing and duration info
+ */
+export const parseByDuration = (start, timeline, timescale, duration, sourceDuration) => {
+  const count = Math.ceil(sourceDuration / (duration / timescale));
+
+  return range(start, start + count).map((number, index) => {
+    const segment = { number, duration: duration / timescale, timeline };
+
+    if (index === count - 1) {
+      // final segment may be less than duration
+      segment.duration = sourceDuration - (segment.duration * index);
+    }
+
+    segment.time = (segment.number - start) * duration;
+
+    return segment;
+  });
+};
+
+/**
+ * Uses information provided by SegmentTemplate.SegmentTimeline to determine segment
+ * timing and duration
+ *
+ * @param {number} start
+ *        The start number for the first segment of this period
+ * @param {number} timeline
+ *        The timeline (period index) for the first segment of this period
+ * @param {number} timescale
+ *        The timescale for the timestamps contained within the media content
+ * @param {Object[]} segmentTimeline
+ *        List of objects representing the attributes of each S element contained within
+ * @param {number} sourceDuration
+ *        Duration of the entire Media Presentation
+ * @return {{number: number, duration: number, time: number, timeline: number}[]}
+ *         List of Objects with segment timing and duration info
+ */
+export const parseByTimeline =
+(start, timeline, timescale, segmentTimeline, sourceDuration) => {
+  const segments = [];
+  let time = -1;
+
+  for (let sIndex = 0; sIndex < segmentTimeline.length; sIndex++) {
+    const S = segmentTimeline[sIndex];
+    const duration = parseInt(S.d, 10);
+    const repeat = parseInt(S.r || 0, 10);
+    const segmentTime = parseInt(S.t || 0, 10);
+
+    if (time < 0) {
+      // first segment
+      time = segmentTime;
+    }
+
+    if (segmentTime && segmentTime > time) {
+      // discontinuity
+
+      // TODO: How to handle this type of discontinuity
```

Comment:
any thoughts on this TODO @forbesjo @gesinger ?

relevant spec text 5.3.9.6
>the signalling of discontinuities of the Media Presentation timeline for which no Segment data are present in a specific Representation.

>If the value of the attribute is greater than what is expressed by the previous S element, it expresses discontinuities in the timeline.

I also found this blog post that mentioned the discontinuities briefly http://www.unified-streaming.com/blog/stop-numbering-underappreciated-power-dashs-segmenttimeline

Based on the wording of the spec and that blog post, this form of discontinuity signaling appears to be the DASH equivalent of EXT-X-GAP

Reply:
Talked about this in chat, we'll leave this as is.